### ISSUE TITLE
capture structured journal meta properties

### ISSUE THREAD

**Issue by woodz-** [rsyslog/rsyslog#5035]
### Expected behavior
capture all meta properties of journal by using RSYSLOG_DebugFormat template

### Actual behavior
no meta properties are being captured

### Steps to reproduce the behavior

I am struggling on how to capture `systemd-journald` properties into `rsyslog` files.

### My setup ###

 - ubuntu inside docker on arm (raspberrypi): `FROM arm64v8/ubuntu:20.04`
 - docker command (all subsequent actions taken inside running docker container)

        $ docker run --privileged -ti --cap-add SYS_ADMIN --security-opt seccomp=unconfined --cgroup-parent=docker.slice --cgroupns private --tmpfs /tmp --tmpfs /run --tmpfs /run/lock systemd:origin

 - `rsyslog` under `$ sytemctl status rsyslog`

        ● rsyslog.service - System Logging Service
             Loaded: loaded (/lib/systemd/system/rsyslog.service; enabled; vendor prese>
             Active: active (running)
        ...
        [origin software="rsyslogd" swVersion="8.2001.0" x-pid="39758" x-info="https://www.rsyslog.com"] start
        ...

### My plan ###

Having a small c program to put some information into journal:

        #include <systemd/sd-journal.h>
        #include <stdio.h>
        #include <unistd.h>

        int main(int arcg, char** args) {
          char buffer [50];
   
          sprintf (buffer, "%lu", (unsigned long)getpid());

          printf("writing to journal\n");
 
          sd_journal_print(LOG_WARNING, "%s", "a little journal test message");
          sd_journal_send("MESSAGE=%s", "there shoud be a text", "SYSLOG_PID=%s", buffer, "PRIORITY=%i", LOG_ERR, "DOCUMENTATION=%s", "any doc link", "MESSAGE_ID=%s", "e5e4132e441541f89bca0cc3e7be3381", "MEAS_VAL=%d", 1394, NULL);

          return 0;
        }

Compile it: `$ gcc joutest.c -lsystemd -o jt`

Execute it: `$ ./jt`

This results inside the journal as `$ journalctl -r -o json-pretty`:

        {
          "_GID" : "0",
          "MESSAGE" : "there shoud be a text",
          "_HOSTNAME" : "f1aad951c039",
          "SYSLOG_IDENTIFIER" : "jt",
          "_TRANSPORT" : "journal",
          "CODE_FILE" : "joutest.c",
          "DOCUMENTATION" : "any doc link",
          "_BOOT_ID" : "06a36b314cee462591c65a2703c8b2ad",
          "CODE_LINE" : "14",
          "MESSAGE_ID" : "e5e4132e441541f89bca0cc3e7be3381",
          "_CAP_EFFECTIVE" : "3fffffffff",
          "__REALTIME_TIMESTAMP" : "1669373862349599",
          "_SYSTEMD_UNIT" : "init.scope",
          "CODE_FUNC" : "main",
          "_MACHINE_ID" : "5aba31746bf244bba6081297fe061445",
          "SYSLOG_PID" : "39740",
          "PRIORITY" : "3",
          "_COMM" : "jt",
          "_SYSTEMD_SLICE" : "-.slice",
          "MEAS_VAL" : "1394",
          "__MONOTONIC_TIMESTAMP" : "390853282189",
          "_PID" : "39740",
          "_SOURCE_REALTIME_TIMESTAMP" : "1669373862336503",
          "_UID" : "0",
          "_SYSTEMD_CGROUP" : "/init.scope",
          "__CURSOR" : "s=63a46a30bbbb4b8c9288a9b12c622b37;i=6cb;b=06a36b314cee46>
        }

Now as a test, extracting all properties from that journal entry via `rsyslog`; property in the jargon of `rsyslog` in principle is the name of a key in the formatted json entry. But if a property (or key name) matches, the whole dictionary item (key and value) shall be captured

To start with this, I've configured `rsyslog` as:

        module(load="imjournal")
        module(load="mmjsonparse")

        action(type="mmjsonparse")

        if $programname == 'jt' and $syslogseverity == 3 then
          action(type="omfile" file="/var/log/jt_err.log" template="RSYSLOG_DebugFormat")

This config is located in `/etc/rsyslog.d/filter.conf` and gets automatically included by `/etc/rsyslog.conf`:

        # /etc/rsyslog.conf configuration file for rsyslog
        #
        # For more information install rsyslog-doc and see
        # /usr/share/doc/rsyslog-doc/html/configuration/index.html
        #
        # Default logging rules can be found in /etc/rsyslog.d/50-default.conf
        
        
        #################
        #### MODULES ####
        #################
        
        #module(load="imuxsock") # provides support for local system logging
        #module(load="immark")  # provides --MARK-- message capability
        
        # provides UDP syslog reception
        #module(load="imudp")
        #input(type="imudp" port="514")
        
        # provides TCP syslog reception
        #module(load="imtcp")
        #input(type="imtcp" port="514")
        
        # provides kernel logging support and enable non-kernel klog messages
        module(load="imklog" permitnonkernelfacility="on")
        
        ###########################
        #### GLOBAL DIRECTIVES ####
        ###########################
        
        #
        # Use traditional timestamp format.
        # To enable high precision timestamps, comment out the following line.
        #
        $ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
        
        # Filter duplicated messages
        $RepeatedMsgReduction on
        
        #
        # Set the default permissions for all log files.
        #
        $FileOwner syslog
        $FileGroup adm
        $FileCreateMode 0640
        $DirCreateMode 0755
        $Umask 0022
        $PrivDropToUser syslog
        $PrivDropToGroup syslog
        
        #
        # Where to place spool and state files
        #
        $WorkDirectory /var/spool/rsyslog
        
        #
        # Include all config files in /etc/rsyslog.d/
        #
        $IncludeConfig /etc/rsyslog.d/*.conf

Applied this config: `$ systemctl restart rsyslog`

Which results in the following: `$ cat /var/log/jt_err.log`

        Debug line with all properties:
        FROMHOST: 'f1aad951c039', fromhost-ip: '127.0.0.1', HOSTNAME: 
        'f1aad951c039', PRI: 11,
        syslogtag 'jt[39765]:', programname: 'jt', APP-NAME: 'jt', PROCID: 
        '39765', MSGID: '-',
        TIMESTAMP: 'Nov 25 11:47:50', STRUCTURED-DATA: '-',
        msg: ' there shoud be a text'
        escaped msg: ' there shoud be a text'
        inputname: imuxsock rawmsg: '<11>Nov 25 11:47:50 jt[39765]: there 
        shoud be a text'
        $!:{ "msg": "there shoud be a text" }
        $.:
        $/:

### My problem ###

Looking on the resulting rsyslog, I miss a majority, if not all, of items originating from the journal entry.
There is really no property (key) matching. Shouldn't be there all properties matched as it is a debug output?

Specifically I am concentrating on my custom property, `MEAS_VAL`, it is not there.
The only property which occurs is "msg", which by the way is questionable whether it is a match of the journal, since the originating property name attached to the resulting content `"there shoud be a text"` is `MESSAGE`

So it feels that I don't hit the whole journal capturing mechanism at all, why?

### Environment
- rsyslog version: 8.2001.0
- platform: Ubuntu 20.04
- for configuration questions/issues, include rsyslog.conf and included config files

**Comment by davidelang** [comment]
according to the message that you logged via RSYSLOG_DebugFormat, that arrived 
to rsyslog via imuxsocket, so either it didn't go through journald at all, or 
this is the copy that journald then delivered to rsyslog /dev/log equivalent, 
after journald decided to throw away everything.

What you need is to find the message that was read in via imjournal instead.

I'm not sure what's going on here, please start rsyslog with -o /path/to/file 
and send us that file (that will be the combined configs, showing everything 
that's included as well as rsyslog.conf), so we can see the complete config.

There may also be something we need to change on the journald side to have it 
NOT deliver the messages to rsyslog via unix sockets. We've tried for YEARS to 
get journald to add an option that would deliver all the properties in JSON 
format to rsyslog, but they refuse to do that because there is no RFC that says 
to do so. So we hvae to disable their default config and instead poll journald 
for new messages to get everything. :-(

David Lang

**Comment by woodz-** [comment]
Many thanks for your reaction to my issue.

Find the requested composition log below:

        
        ## full conf created by rsyslog version 8.2001.0 at 2022-11-28 08:48:04 ##
        
        ##### BEGIN CONFIG: /etc/rsyslog.conf
        # /etc/rsyslog.conf configuration file for rsyslog
        #
        # For more information install rsyslog-doc and see
        # /usr/share/doc/rsyslog-doc/html/configuration/index.html
        #
        # Default logging rules can be found in /etc/rsyslog.d/50-default.conf
        
        
        #################
        #### MODULES ####
        #################
        
        module(load="imuxsock") # provides support for local system logging
        #module(load="immark")  # provides --MARK-- message capability
        
        # provides UDP syslog reception
        #module(load="imudp")
        #input(type="imudp" port="514")
        
        # provides TCP syslog reception
        #module(load="imtcp")
        #input(type="imtcp" port="514")
        
        # provides kernel logging support and enable non-kernel klog messages
        module(load="imklog" permitnonkernelfacility="on")
        
        ###########################
        #### GLOBAL DIRECTIVES ####
        ###########################
        
        #
        # Use traditional timestamp format.
        # To enable high precision timestamps, comment out the following line.
        #
        $ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
        
        # Filter duplicated messages
        $RepeatedMsgReduction on
        
        #
        # Set the default permissions for all log files.
        #
        $FileOwner syslog
        $FileGroup adm
        $FileCreateMode 0640
        $DirCreateMode 0755
        $Umask 0022
        $PrivDropToUser syslog
        $PrivDropToGroup syslog
        
        #
        # Where to place spool and state files
        #
        $WorkDirectory /var/spool/rsyslog
        
        #
        # Include all config files in /etc/rsyslog.d/
        #
        $IncludeConfig /etc/rsyslog.d/*.conf/etc/rsyslog.d/*.conf
        ##### BEGIN CONFIG: /etc/rsyslog.d/filter.conf
        
        ##### BEGIN CONFIG: /etc/rsyslog.d/50-default.conf
        #  Default rules for rsyslog.
        #
        #			For more information see rsyslog.conf(5) and /etc/rsyslog.conf
        
        #
        # First some standard log files.  Log by facility.
        #
        auth,authpriv.*			/var/log/auth.log
        *.*;auth,authpriv.none		-/var/log/syslog
        #cron.*				/var/log/cron.log
        #daemon.*			-/var/log/daemon.log
        kern.*				-/var/log/kern.log
        #lpr.*				-/var/log/lpr.log
        mail.*				-/var/log/mail.log
        #user.*				-/var/log/user.log
        
        #
        # Logging for the mail system.  Split it up so that
        # it is easy to write scripts to parse these files.
        #
        #mail.info			-/var/log/mail.info
        #mail.warn			-/var/log/mail.warn
        mail.err			/var/log/mail.err
        
        #
        # Some "catch-all" log files.
        #
        #*.=debug;#	auth,authpriv.none;#	news.none;mail.none	-/var/log/debug
        PreprocFileLineNumber(31)
        #*.=info;*.=notice;*.=warn;#	auth,authpriv.none;#	cron,daemon.none;#	mail,news.none		-/var/log/messages
        PreprocFileLineNumber(35)
        
        #
        # Emergencies are sent to everybody logged in.
        #
        *.emerg				:omusrmsg:*
        
        #
        # I like to have messages displayed on the console, but only on a virtual
        # console I usually leave idle.
        #
        #daemon,mail.*;#	news.=crit;news.=err;news.=notice;#	*.=debug;*.=info;#	*.=notice;*.=warn	/dev/tty8
        
        ##### END   CONFIG: /etc/rsyslog.d/50-default.conf
        #filter testing rsyslog
        #if programname is jt and syslogseverity is err (3) -> log to /var/log/jt_err.log
        
        module(load="imjournal")
        module(load="mmjsonparse")
        
        action(type="mmjsonparse")
        
        
        if $programname == 'jt' and $syslogseverity == 3 then
          action(type="omfile" file="/var/log/jt_err.log" template="RSYSLOG_DebugFormat")
        
        
        
        ##### END   CONFIG: /etc/rsyslog.d/filter.conf
        
        
        ##### END   CONFIG: /etc/rsyslog.conf
        
I got a little confused why you say my message gets captured through `imuxsock`. But now I understand.

I had intentionally commented the `module(load="imuxsock")` and then received no debug log message at all. Only after uncommenting I got the same result back as reported.

And of course the debug log tells that the message comes from `imuxsock`.

**Comment by woodz-** [comment]
Meanwhile I've compiled my own version `8.2212.0.master`. But the phenomenon persists.

**Comment by woodz-** [comment]
Could these `systemd's` startup messages indicate the problem? Will the position remain on head after logging with my sample program?

        Nov 28 16:27:38 f1aad951c039 rsyslogd[144703]: imjournal: Journal indicates no msgs when positioned at head.  [v8.2212.0.master try https://www.rsyslog.com/e/0 ]
        Nov 28 16:27:38 f1aad951c039 rsyslogd[144703]: imjournal: journal files changed, reloading...  [v8.2212.0.master try https://www.rsyslog.com/e/0 ]
        Nov 28 16:27:38 f1aad951c039 rsyslogd[144703]: imjournal: Journal indicates no msgs when positioned at head.  [v8.2212.0.master try https://www.rsyslog.com/e/0 ]

**Comment by woodz-** [comment]
Isn't there a debug log, for `imjournal` and `mmjsonparse` modules to see what they try to do and on which point they may be fail?

**Comment by rgerhards** [comment]
Apologies for the very late reply. This is probably no longer relevant to the original reporter, but I am adding the conclusion for people who arrive here via search while we work through the backlog. We do not close rsyslog issues lightly, and we are not closing this merely because of age.

Triage result: closed as answered with imjournal structured-data/template guidance.

Reasoning: The current imjournal documentation confirms that structured journal data requires imjournal, and the supported output pattern is a template using %$!all-json% rather than RSYSLOG_DebugFormat alone.

If this still reproduces on a current rsyslog release and maintained platform, please open a new issue with the current version, complete configuration, logs, and a minimal reproducer so we can investigate it as a fresh current problem.
